src/pymor/operators/constructions.py

```diff
@@ -1430,3 +1430,81 @@ def apply_adjoint(self, V, mu=None):
     def apply_inverse_adjoint(self, U, mu=None, initial_guess=None, least_squares=False):
         assert U in self.source
         return self.range.from_numpy(U.to_numpy())
+
+
+class BilinearFunctional(Operator):
+    """An `Operator` representing a bilinear functional.
+
+    Given an operator `A` acting on R^d, this class enables the functional::
+
+        op: R^d ----> R^1
+             u  |---> A(u, u)
+
+    In particular this is intended to provide the jacobian of such a bilinear
+    functional::
+
+        op.jacobian(U, mu).apply(V) = (A(u, v; mu) + A(v, u; mu))
+
+    Parameters
+    ----------
+    operator
+        The `Operator` to be wrapped as a bilinear functional.
+    """
+
+    linear = False
+    range = NumpyVectorSpace(1)
+
+    def __init__(self, operator, name=None):
+        assert operator.linear
+        assert operator.source == operator.range
+        self.__auto_init(locals())
+        self.source = operator.source
+
+    def apply(self, U, mu=None):
+        assert U in self.source
+        return self.range.from_numpy(self.operator.apply2(U, U, mu))
+
+    def jacobian(self, U, mu=None):
+        inner_vec = self.operator.apply_adjoint(U, mu) + self.operator.apply(U, mu)
+        return VectorFunctional(inner_vec, name=self.name + '_jacobian')
+
+    def d_mu(self, parameter, index=1):
+        # the parameter derivative only takes effect on the inner operator
+        return BilinearFunctional(self.operator.d_mu(parameter, index), name=self.name + '_d_mu')
+
+
+class BilinearProductFunctional(BilinearFunctional):
+    """An `Operator` representing a bilinear functional by multiplying two linear functionals.
+
+    Given linear operators `A`, `B`: R^d ----> R^n, this class enables the functional::
+
+        op: R^d ----> R^1
+             u  |---> (A(u), B(u))
+
+    In particular this is intended to provide the jacobian of such a bilinear
+    functional::
+
+        op.jacobian(U, mu).apply(V) = (A(v; mu), B(u; mu)) + (A(u; mu), B(v; mu))
+
+    Parameters
+    ----------
+    operators
+        The `Operator`s to be wrapped as a bilinear functional.
+    """
+
+    linear = False
+    range = NumpyVectorSpace(1)
+
+    def __init__(self, operators, product=None, name=None):
```

Comment:
I would expect parameters `A` and `B` instead of `operators`. Any reason for requiring to pass the operators as a pair?

Reply:
I wanted to keep the signature similar to `BilinearFunctional`. I'm going to take a page from the `LowRankOperator` book and call them left and right instead of `A` and `B`, if that's fine with you.